File: C(++)/002. Add Two Numbers.c
/*
002. Add Two Numbers
https://leetcode.com/problems/add-two-numbers/

c 19 ms
*/

/*
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode *pt=(struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *p=pt;
    p->val=0;
    p->next=NULL;
    int d,stp=0;
    while (true)
    {
        if (l1==NULL) d=0;
        else
        {
            d=l1->val;
            l1=l1->next;
        }
        if (l2==NULL) d+=0;
        else
        {
            d+=l2->val;
            l2=l2->next;
        }
        p->val=d+stp;
        stp=0;
        if (p->val>9)
        {
            stp=1;
            p->val-=10;
        }
        if (l1==NULL&&l2==NULL&&stp==0) break;
        else
        {
            p->next=(struct ListNode *)malloc(sizeof(struct ListNode));
            p=p->next;
            p->val=0;
            p->next=NULL;
        }
    }
    return pt;
}

File: Java/002. Add Two Numbers.java
/*
002. Add Two Numbers
https://leetcode.com/problems/add-two-numbers/

Java 73 ms
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1==null&&l2==null) return null;
        
        ListNode ret=null,temp=null;
        int val1,val2,r=0;
        while (l1!=null||l2!=null)
        {
        	if (ret==null)
        	{
        		ret=new ListNode(0);
        		temp=ret;
        	}
        	else
        	{
        		// =temp.next放在后边，因为next还不存在
                temp.next=new ListNode(0);
                temp=temp.next;
        	}
        	
            if (l1==null) val1=0;
            else val1=l1.val;
            if (l2==null) val2=0;
            else val2=l2.val;
            
            temp.val=(val1+val2+r)%10;
            r=(val1+val2+r)/10;
            
            if (l1!=null) l1=l1.next;
            if (l2!=null) l2=l2.next;
        }
        if (r!=0)
        {
            temp.next=new ListNode(r);
            temp=temp.next;
        }
        
        return ret;
    }
}
